### DISCUSSION TITLE
Configuring renovate against private nexus required non-obvious registry URL

### DISCUSSION THREAD

**Discussion by annappropriate** [renovatebot/renovate#18029] [Suggest an Idea]
Hi! I want to document a solution for a problem that we hit when setting up renovate with a private maven repo in nexus. The project is sbt, however, I am certain that the issue can arise regardless of the manager. We have run renovate using both npx and docker with the same results, version 32.208.0.
Summary of the issue is: when using a dependency published to a private maven repository in nexus, we were getting "package not found" error with no logs or anything else pointing to why that is happening. Our sbt setup is such that we have a custom plugin that all projects use, and the resolvers are specified in the plugin, so renovate could not access the URLs (they are defined in a different repository). To mitigate that, we decided to add a renovate package rule that would direct all our internal packages to be resolved from a specified repository, so that we don't have to duplicate the resolvers in every project. Here's a minimal example, omitting the plugin code.
`build.sbt`
```scala
libraryDependencies += "com.ourcompany" %% "ourpackage" % "1.2.3"
enablePlugins(OurPlugin)
```
`project/plugins.sbt`
```scala
resolvers ++= Seq("Plugin releases" at "https://nexus.ourcompany.net/repository/plugin-releases/")
credentials += Credentials(Path.userHome / ".sbt" / ".credentials")
addSbtPlugin("com.ourcompany" % "sbt-plugin" % "5.6.7")
```
`renovate.js` (supplied as config file to renovate)
```js
module.exports = {
    hostRules: [
        {
            matchHost: "nexus.ourcompany.net",
            username: process.env.NEXUS_USER,
            password: process.env.NEXUS_PASS
        }
    ],
    packageRules: [
        {
            matchDatasources: ["sbt-package"],
            matchPackagePrefixes: ["com.ourcompany"],
            registryUrls: ["https://nexus.ourcompany.net/repository/maven-releases/"],
        },
    ]
}
```
We are using the registry URL specified above as a resolver in the plugin configuration, and all sbt projects can resolve it fine as a dependency. However, this resulted in private packages not resolved in renovate.

After a lot of digging, it turned out that when the registry URL was downloaded as the index page here https://github.com/renovatebot/renovate/blob/9caf45ed435e418924edaa93138414eb95b7f484/lib/modules/datasource/sbt-package/index.ts#L39-L42, it contained not an HTML page with package list (like renovate expected), but rather a page that looks like this:
<details>
<summary>Click to view HTML page</summary>

```html
$ curl -u $NEXUS_USER:$NEXUS_PASS https://nexus.ourcompany.net/repository/maven-releases/

<!DOCTYPE html>
<html lang="en">
<head>
  <title>Repository - Nexus Repository Manager</title>
  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>


  <!--[if lt IE 9]>
  <script>(new Image).src="../../favicon.ico?3.42.0-01"</script>
  <![endif]-->
  <link rel="icon" type="image/png" href="../../favicon-32x32.png?3.42.0-01" sizes="32x32">
  <link rel="mask-icon" href="../../safari-pinned-tab.svg?3.42.0-01" color="#5bbad5">
  <link rel="icon" type="image/png" href="../../favicon-16x16.png?3.42.0-01" sizes="16x16">
  <link rel="shortcut icon" href="../../favicon.ico?3.42.0-01">
  <meta name="msapplication-TileImage" content="../../mstile-144x144.png?3.42.0-01">
  <meta name="msapplication-TileColor" content="#00a300">

  <link rel="stylesheet" type="text/css" href="../../static/css/nexus-content.css?3.42.0-01"/>
</head>
<body>
<div class="nexus-header">
  <a href="../..">
    <div class="product-logo">
      <img src="../../static/rapture/resources/icons/x32/nexus-white.png?3.42.0-01" alt="Product logo"/>
    </div>
    <div class="product-id">
      <div class="product-id__line-1">
        <span class="product-name">Nexus Repository Manager</span>
      </div>
      <div class="product-id__line-2">
        <span class="product-spec">OSS 3.42.0-01</span>
      </div>
    </div>
  </a>
</div>

<div class="nexus-body">
  <div class="content-header">
    <img src="../../static/rapture/resources/icons/x32/database.png?3.42.0-01" alt="Repository image"/>
    <span class="title">Repository</span>
    <span class="description">maven-releases</span>
  </div>

  <div class="content-body">
    <div class="content-section">
      <p>
        This maven2 hosted repository is not directly browseable at this URL.
      </p>

      <p>
        Please use the <a href="../../#browse/browse:maven-releases">browse</a>
        or <a href="../../service/rest/repository/browse/maven-releases/">HTML index</a>
        views to inspect the contents of this repository.
      </p>
    </div>
  </div>
</body>
</html>
```
</details>

After that, the parser was extracting the `../../service/rest/repository/browse/maven-releases/` URL, going to this return statement and returning `false` after expression evaluation https://github.com/renovatebot/renovate/blob/9caf45ed435e418924edaa93138414eb95b7f484/lib/modules/datasource/sbt-package/index.ts#L55

Basically, our nexus was not providing a nice index of all packages, but rather offering a human-readable redirect. 

The overall solution was to modify the `packageRules.registryUrls` to be `["https://nexus.ourcompany.net/service/rest/repository/browse/maven-releases/"]`, essentially resolving the redirect manually.
With this change, renovate is able to resolve private packages successfully.

I see two options here:
1. Try to apply the suggested URL modification to the registry URL, since it's possible to identify that a modification is necessary by the ".." prefix
2. Add some logs stating that a URL does not seem valid and/or that the index page does not look right to renovate.

I can consider making a PR, however I would appreciate getting input from the developers in such a case. I could also add a full reproduction example if needed, however that would require setting up multiple repos and a nexus docker image.

**Comment by viceice** [comment]
btw: `authType` is invalid on your host rule, checkout the docs for right usage

**Reply by annappropriate** [reply]
Thanks, I'll take a look

**Reply by annappropriate** [reply]
You're right, it's pointless, removed it now.

**Comment by pk27734** [comment]
I'm encountering an issue with a private Nexus as well:

```
DEBUG: Content is not found for Maven url (repository=dev-team-1/example/parent-pom)
       "url": "https://nexus.company.internal/repository/browse/maven-releases/example/commons/maven-metadata.xml",
       "statusCode": undefined
```

The URL `https://nexus.company.internal/repository/browse/maven-releases/example/commons/maven-metadata.xml` doesn't work, but without _browse_ it does: `https://nexus.company.internal/repository/maven-releases/example/commons/maven-metadata.xml`

How can I get this to work? Setting either option in the `packageRules.registryUrls` doesn't make a difference
